### PR TITLE
feat: markdown config typechecking

### PR DESCRIPTION
**Pull request by JuanM04**
## Changes

Following the comment `// TODO: add better type checking`, I've added type checking for the markdown configuration

### CODE REVIEW

File: packages/astro/src/@types/astro.ts
Comment (FredKSchott) [inline packages/astro/src/@types/astro.ts:352]: can you change to `@typeraw`? I think this would error as invalid JSDoc

File: packages/astro/src/core/config.ts
Comment (FredKSchott) [inline packages/astro/src/core/config.ts:16]: wahhh??? Would love an explanation why here, for future readers

File: packages/astro/src/core/config.ts
Comment (FredKSchott) [inline packages/astro/src/core/config.ts:145]: only need 1 `default()` call

File: packages/astro/src/core/config.ts
Comment (JuanM04) [reply packages/astro/src/core/config.ts:17]: Check out the new commit, I added a longer (not sure if better) explanation

File: packages/astro/src/@types/astro.ts
Comment (natemoo-re) [inline packages/astro/src/@types/astro.ts:361]: I don't think this actually does anything right now, which is why we didn't include it!

**Review by natemoo-re — CHANGES_REQUESTED** [review]
I'd remove any public references to the `mode` option for now, since it doesn't do anything.

Other than that, looks great!

File: packages/astro/src/@types/astro.ts
Comment (JuanM04) [reply packages/astro/src/@types/astro.ts:361]: It does add `rehypeJsx` and `rehypeExpressions`. I could remove the MDX mode altogether if these plugins do nothing

Comment (natemoo-re) [reply packages/astro/src/@types/astro.ts:361]: Ah no you're right! I also saw that https://github.com/withastro/astro/pull/2971 added escaping for expressions when `mode === 'md'`. So I think this is fine!

**Review by natemoo-re — APPROVED** [review]

File: packages/astro/src/@types/astro.ts
Comment (JuanM04) [reply packages/astro/src/@types/astro.ts:361]: After doing some testing, the only difference I've found is that components are being rendered with `mdx` but completely ignored with `md`. Expressions works in both modes, tho.

**Review by FredKSchott — CHANGES_REQUESTED** [review]
Marking as red just to make sure we hold off on this until post release day. I have some comments that I'd like to make but just won't have the time until Monday/Tuesday to make them. Appreciate the patience!

File: packages/astro/src/core/config.ts
Comment (FredKSchott) [inline packages/astro/src/core/config.ts:150]: Thanks again for your patience on this review @JuanM04! 

My main comment is that we really benefit from having all of our config parsing logic here in one file and in one single parser object instantiation. 

Reading this PR, I can see the value you've put into validation reuse, which is great. But, I'm also bouncing around to follow where the different parsers / validation logic lives, defined across multiple files, and even multiple packages, with multiple zod parser instances referencing each other.

If you'd like to improve the existing `markdown: z.object(...` validation logic defined here, I'm all for it. I could even see the value in a new `Plugin` validator that both `remarkPlugins` and `rehypePlugins`, which breaks up the single-object-validator that I love but for a valid, valuable reason.

But at the end of the day, Astro owns the "Astro Config" type and therefore its validation, so I'd like to keep it as simple as possible and only inside the Astro package, for our sake as maintainers.

@JuanM04 lmk what you think, and how you'd like to move forward!

**Review by FredKSchott — CHANGES_REQUESTED** [review]

File: packages/astro/src/core/config.ts
Comment (JuanM04) [reply packages/astro/src/core/config.ts:150]: You're totally right. The idea behind the division was about separation of concerns, since `@astrojs/markdown-remark` is its own package, but I see how that can be a problem.

I'll integrate this configuration in a single place

**Comment by JuanM04** [comment]
@FredKSchott I've moved the validation inside `config.ts`. Also, I've removed all the internal references to the old `markdownRender` way of importing markdown, so it's cleaner and doesn't do unnecessary work

**Comment by JuanM04** [comment]
Note: the error thrown by Deno it's caused by this PR directly. If you have an Astro page with the `<Markdown />` component and build it with Deno, it will throw "No Markdown parser found"

**Review by FredKSchott — APPROVED** [review]
Nice! This is looking great, thanks for making those changes. LGTM!

**Comment by FredKSchott** [comment]
> Note: the error thrown by Deno it's not caused by this PR directly. If you have an Astro page with the `<Markdown />` component and build it with Deno, it will throw "No Markdown parser found"

Can't merge with failing tests! What's the best way to resolve?

**Comment by JuanM04** [comment]
> Can't merge with failing tests! What's the best way to resolve?

I'm looking into it. Either a fix is found or the test is ignored

**Comment by FredKSchott** [comment]
As long as you can confirm that this isn't a regression caused by this PR

**Comment by JuanM04** [comment]
Ok, this way it should work. The plugin now is dynamically imported only when using `<Markdown />` _and_ Node.js. When using Deno, it throws an error

**Review by FredKSchott — APPROVED** [review]
LGTM!

**Pull request merged**
